File: lib/src/test/java/com/hammwerk/wizardpager/core/BranchTest.java
package com.hammwerk.wizardpager.core;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import de.bechte.junit.runners.context.HierarchicalContextRunner;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNull;

@RunWith(HierarchicalContextRunner.class)
public class BranchTest {
	@Before
	public void setUp() throws Exception {
	}

	@Test
	public void createBranch() throws Exception {
		Branch branch = new Branch();
		assertNull(branch.getName());
	}

	@Test
	public void createBranchWithName() throws Exception {
		Branch branch = new Branch("Name");
		assertEquals("Name", branch.getName());
	}

	public class GivenABranch {
		private Branch branch;

		@Before
		public void setUp() throws Exception {
			branch = new Branch();
		}

		@Test
		public void whenSetPages_thenReturnBranch() throws Exception {
			assertEquals(branch, branch.setPages());
		}

		@Test(expected = Branch.PageAfterBranchPageException.class)
		public void whenSetPagesWithPageAfterBranchPage_thenThrowPageAfterBranchPageException() throws Exception {
			branch.setPages(new TestBranchPage("Branch Page"), new TestPage("Page"));
		}

		@Test(expected = PageIndexOutOfBoundsException.class)
		public void whenGetPage_thenThrowPageIndexOutOfBoundsException() throws Exception {
			branch.getPage(0);
		}

		@Test
		public void whenGetNumberOfPages_thenReturnZero() throws Exception {
			assertEquals(0, branch.getNumberOfPages());
		}

		@Test
		public void whenGetBranchPage_thenReturnNull() throws Exception {
			assertNull(branch.getBranchPage());
		}

		@Test
		public void whenGetNumberOfValidPages_thenReturnZero() throws Exception {
			assertEquals(0, branch.getNumberOfValidPages());
		}

		public class GivenTowPages {
			private Page pageInFirstBranch;
			private Page pageInSecondBranch;

			@Before
			public void setUp() throws Exception {
				pageInFirstBranch = new TestPage("Page in first Branch");
				pageInSecondBranch = new TestPage("Page in second Branch");
				branch.setPages(pageInFirstBranch, pageInSecondBranch);
			}

			@Test
			public void whenGetFirstPage_thenReturnFirstPage() throws Exception {
				assertEquals(pageInFirstBranch, branch.getPage(0));
			}

			@Test(expected = PageIndexOutOfBoundsException.class)
			public void whenGetPageAfterLastPage_thenThrowPageIndexOutOfBoundsException() throws Exception {
				branch.getPage(2);
			}

			@Test
			public void whenGetBranchPage_thenReturnNull() throws Exception {
				assertNull(branch.getBranchPage());
			}

			@Test
			public void whenGetNumberOfValidPages_thenReturnTwo() throws Exception {
				assertEquals(2, branch.getNumberOfValidPages());
			}
		}

		public class GivenTwoPagesAndBranchPage {
			private BranchPage branchPage;
			private Page pageInFirstBranch;
			private Page pageInSecondBranch;

			@Before
			public void setUp() throws Exception {
				pageInFirstBranch = new TestPage("Page in first Branch");
				pageInSecondBranch = new TestPage("Page in second Branch");
				branchPage = new TestBranchPage("Title")
						.addBranch("First Branch", pageInFirstBranch)
						.addBranch("Second Branch", pageInSecondBranch);
				branch.setPages(pageInFirstBranch, pageInSecondBranch, branchPage);
			}

			@Test
			public void whenGetNumberOfPages_thenReturnNumberOfPages() throws Exception {
				assertEquals(3, branch.getNumberOfPages());
			}

			@Test
			public void whenGetBranchPage_thenReturnBranchPage() throws Exception {
				assertEquals(branchPage, branch.getBranchPage());
			}
		}

		public class GivenARequiredPage {
			private Page requiredPage;

			@Before
			public void setUp() throws Exception {
				requiredPage = new TestPage("Required Page", true);
				branch.setPages(requiredPage);
			}

			@Test
			public void whenGetNumberOfValidPages_thenReturnZero() throws Exception {
				assertEquals(0, branch.getNumberOfValidPages());
			}

			public class GivenCompletedRequiredPage {
				@Before
				public void setUp() throws Exception {
					requiredPage.setCompleted();
				}

				@Test
				public void whenGetNumberOfValidPages_thenReturnOne() throws Exception {
					assertEquals(1, branch.getNumberOfValidPages());
				}
			}
		}

		public class GivenTwoRequiredPages {
			private Page firstRequiredPage;
			private Page secondRequiredPage;

			@Before
			public void setUp() throws Exception {
				firstRequiredPage = new TestPage("First required Page", true);
				secondRequiredPage = new TestPage("Second required Page", true);
				branch.setPages(firstRequiredPage, secondRequiredPage);
			}

			@Test
			public void whenGetNumberOfValidPages_thenReturnOne() throws Exception {
				assertEquals(0, branch.getNumberOfValidPages());
			}

			public class GivenCompletedFirstRequiredPage {
				@Before
				public void setUp() throws Exception {
					firstRequiredPage.setCompleted();
				}

				@Test
				public void whenGetNumberOfValidPages_thenReturnOne() throws Exception {
					assertEquals(1, branch.getNumberOfValidPages());
				}
			}

			public class GivenCompletedAllRequiredPage {
				@Before
				public void setUp() throws Exception {
					firstRequiredPage.setCompleted();
					secondRequiredPage.setCompleted();
				}

				@Test
				public void whenGetNumberOfValidPages_thenReturnOne() throws Exception {
					assertEquals(2, branch.getNumberOfValidPages());
				}
			}
		}
	}
}

File: lib/src/test/kotlin/com/hammwerk/wizardpager/core/BranchTest.kt
package com.hammwerk.wizardpager.core

import de.bechte.junit.runners.context.HierarchicalContextRunner
import junit.framework.Assert.assertEquals
import junit.framework.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@Suppress("UNUSED")
@RunWith(HierarchicalContextRunner::class)
class BranchTest {
	inner class GivenABranch {
		private lateinit var branch: Branch

		@Before
		fun setUp() {
			branch = Branch()
		}

		@Test
		fun whenGetName_thenReturnNull() {
			assertNull(branch.name)
		}

		@Test(expected = Branch.PageAfterBranchPageException::class)
		fun whenSetPagesWithPageAfterBranchPage_thenThrowPageAfterBranchPageException() {
			branch.pages = arrayOf(BranchPage("Branch Page", "FragmentName"), TestPage("Page"))
		}

		@Test(expected = PageIndexOutOfBoundsException::class)
		fun whenGetPage_thenThrowPageIndexOutOfBoundsException() {
			branch.getPage<Page>(0)
		}

		@Test
		fun whenGetNumberOfPages_thenReturnZero() {
			assertEquals(0, branch.numberOfPages)
		}

		@Test
		fun whenGetBranchPage_thenReturnNull() {
			assertNull(branch.branchPage)
		}

		@Test
		fun whenGetNumberOfValidPages_thenReturnZero() {
			assertEquals(0, branch.numberOfValidPages)
		}

		inner class GivenTowPages {
			private lateinit var firstPage: Page
			private lateinit var secondPage: Page

			@Before
			fun setUp() {
				firstPage = TestPage("First Page")
				secondPage = TestPage("Second Page")
				branch.pages = arrayOf(firstPage, secondPage)
			}

			@Test
			fun whenGetFirstPage_thenReturnFirstPage() {
				assertEquals(firstPage, branch.getPage<Page>(0))
			}

			@Test(expected = PageIndexOutOfBoundsException::class)
			fun whenGetPageAfterLastPage_thenThrowPageIndexOutOfBoundsException() {
				branch.getPage<Page>(2)
			}

			@Test
			fun whenGetBranchPage_thenReturnNull() {
				assertNull(branch.branchPage)
			}

			@Test
			fun whenGetNumberOfValidPages_thenReturnTwo() {
				assertEquals(2, branch.numberOfValidPages)
			}
		}

		inner class GivenTwoPagesAndBranchPage {
			private lateinit var firstPage: Page
			private lateinit var secondPage: Page
			private lateinit var branchPage: BranchPage
			private lateinit var pageInFirstBranch: Page
			private lateinit var pageInSecondBranch: Page

			@Before
			fun setUp() {
				firstPage = TestPage("First Page")
				secondPage = TestPage("Second Page")
				pageInFirstBranch = TestPage("Page in first Branch")
				pageInSecondBranch = TestPage("Page in second Branch")
				branchPage = BranchPage("Title", "FragmentName").apply {
					addBranch("First Branch", pageInFirstBranch)
					addBranch("Second Branch", pageInSecondBranch)
				}
				branch.pages = arrayOf(firstPage, secondPage, branchPage)
			}

			@Test
			fun whenGetNumberOfPages_thenReturnNumberOfPages() {
				assertEquals(3, branch.numberOfPages)
			}

			@Test
			fun whenGetBranchPage_thenReturnBranchPage() {
				assertEquals(branchPage, branch.branchPage)
			}
		}

		inner class GivenARequiredPage {
			private lateinit var requiredPage: Page

			@Before
			fun setUp() {
				requiredPage = TestPage("Required Page", true)
				branch.pages = arrayOf(requiredPage)
			}

			@Test
			fun whenGetNumberOfValidPages_thenReturnZero() {
				assertEquals(0, branch.numberOfValidPages)
			}

			inner class GivenCompletedRequiredPage {
				@Before
				fun setUp() {
					requiredPage.completed = true
				}

				@Test
				fun whenGetNumberOfValidPages_thenReturnOne() {
					assertEquals(1, branch.numberOfValidPages)
				}
			}
		}

		inner class GivenTwoRequiredPages {
			private lateinit var firstRequiredPage: Page
			private lateinit var secondRequiredPage: Page

			@Before
			fun setUp() {
				firstRequiredPage = TestPage("First required Page", true)
				secondRequiredPage = TestPage("Second required Page", true)
				branch.pages = arrayOf(firstRequiredPage, secondRequiredPage)
			}

			@Test
			fun whenGetNumberOfValidPages_thenReturnOne() {
				assertEquals(0, branch.numberOfValidPages)
			}

			inner class GivenCompletedFirstRequiredPage {
				@Before
				fun setUp() {
					firstRequiredPage.completed = true
				}

				@Test
				fun whenGetNumberOfValidPages_thenReturnOne() {
					assertEquals(1, branch.numberOfValidPages)
				}
			}

			inner class GivenCompletedAllRequiredPage {
				@Before
				fun setUp() {
					firstRequiredPage.completed = true
					secondRequiredPage.completed = true
				}

				@Test
				fun whenGetNumberOfValidPages_thenReturnOne() {
					assertEquals(2, branch.numberOfValidPages)
				}
			}
		}
	}

	inner class GivenABranchWithName {
		private lateinit var branchWithName: Branch

		@Before
		fun setUp() {
			branchWithName = Branch("Name")
		}

		@Test
		fun whenGetName_thenReturnName() {
			assertEquals("Name", branchWithName.name)
		}
	}
}
